### PR TITLE
Add a hostname argument to DNS-related logger methods

### PR DESCRIPTION
**Pull request by AdamStepan**

### CODE REVIEW

File: packages/_debian-common/changelog [hunk 1]
```diff
@@ -1,9 +1,12 @@
-libsuperiormysqlpp (0.3.3) UNRELEASED; urgency=medium
+libsuperiormysqlpp (0.4.0) unstable; urgency=medium
```
Comment (MrWarlockX) [inline packages/_debian-common/changelog]: I would leave UNRELEASED here, if you take look at master commits, you will find release commits, that changes this line.

Comment (AdamStepan) [reply packages/_debian-common/changelog]: ACK

File: tests-extended/dnsa_connection_pool.cpp [hunk 2]
```diff
@@ -192,5 +192,46 @@ go_bandit([]() {
             std::this_thread::sleep_for(10ms);
             AssertThat(pool.poolState().available < min_spare_connections, IsTrue());
         });
+
+        it("can log dns error", [&]() {
+
+            auto& settings = getSettingsRef();
+            std::string hostname{"definitiely.not.resolvable.hostname"}, streamRedirectionResult{};
```
Comment (OpatrilPeter) [inline tests-extended/dnsa_connection_pool.cpp]: streamRedirectionResult nepotrebuje `{}`

File: tests-extended/dnsa_connection_pool.cpp [hunk 3]
```diff
@@ -192,5 +192,46 @@ go_bandit([]() {
             std::this_thread::sleep_for(10ms);
             AssertThat(pool.poolState().available < min_spare_connections, IsTrue());
         });
+
+        it("can log dns error", [&]() {
+
+            auto& settings = getSettingsRef();
```
Comment (OpatrilPeter) [inline tests-extended/dnsa_connection_pool.cpp]: Mas tady i na par jinych mistech nekonzistentni umisteni `&`, doporucuju sjednotit doprava.

File: tests-extended/makefile [hunk 4]
```diff
@@ -1,6 +1,8 @@
 #!/usr/bin/make -f
 
 LDFLAGS += -lboost_system
+# XXX: can be removed when we delete deprecated DNS logger methods
```
Comment (OpatrilPeter) [inline tests-extended/makefile:4]: jestli tu neni XXX uz nekde pouzivane, radsi bych videl TODO

File: tests/db_access/test_utils.hpp [hunk 5]
```diff
@@ -7,7 +7,34 @@
...
 #include <thread>
+#include <iostream>
+#include <streambuf>
+#include <string>
+#include <sstream>
 
+template <typename T>
+class StreamRedirect {
+    public:
+        StreamRedirect(T &stream): buffer{}, originalStream{stream},
```
Comment (OpatrilPeter) [inline tests/db_access/test_utils.hpp]: neni ten buffer{} zbytecny?

File: include/superior_mysqlpp/logging.hpp [hunk 6]
```diff
@@ -618,53 +637,61 @@ namespace SuperiorMySqlpp {
...
+            [[deprecated]] virtual void logSharedPtrPoolDnsAwarePoolManagementStopped(std::uint_fast64_t /*poolId*/)const noexcept override {}
+            [[deprecated]] virtual void logSharedPtrPoolDnsAwarePoolManagementError(std::uint_fast64_t /*poolId*/, std::exception&) const noexcept override {}
+            [[deprecated]] virtual void logSharedPtrPoolDnsAwarePoolManagementError(std::uint_fast64_t /*poolId*/) const noexcept override {}
 
-            virtual void logSharedPtrPoolDnsAwarePoolManagementCycleStart(std::uint_fast64_t poolId) const noexcept override
+            virtual void logSharedPtrPoolDnsAwarePoolManagementCycleStart(std::uint_fast64_t poolId, const std::string &hostname) const noexcept override
             {
                 SpinGuard guard{lock};
-                std::cerr << stringify("Pool [", poolId, "]: DnsAwarePoolManagement: Cycle start.\n") << std::flush;
+                std::cerr << stringify("Pool [", poolId, ", hostname=", hostname, "]: DnsAwarePoolManagement: Cycle start.\n") << std::flush;
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/logging.hpp:655]: mozna je to out-of-scope, ale co bys rikal na to, ze bychom se vyhnuli tomu vytvareni temporary stringu a predelali to na primy vypis do streamu? Ten stringify_helper je genericky a sel by pouzit, takze by to chtelo jen pridat funkci `toStream` co bere stream a varargs a vraci dany stream (abys mohl udelat ten flush)

File: include/superior_mysqlpp/logging.hpp [hunk 7]
```diff
@@ -212,14 +222,23 @@ namespace SuperiorMySqlpp {
...
-            virtual void logSharedPtrPoolDnsAwarePoolManagementChangeDetected(std::uint_fast64_t /*poolId*/) const noexcept override {}
-            virtual void logSharedPtrPoolDnsAwarePoolManagementCheckError(std::uint_fast64_t /*poolId*/, std::exception&) const noexcept override {}
-            virtual void logSharedPtrPoolDnsAwarePoolManagementCheckError(std::uint_fast64_t /*poolId*/) const noexcept override {}
-            virtual void logSharedPtrPoolDnsAwarePoolManagementCycleEnd(std::uint_fast64_t /*poolId*/) const noexcept override {}
-            virtual void logSharedPtrPoolDnsAwarePoolManagementStopped(std::uint_fast64_t /*poolId*/) const noexcept override {}
-            virtual void logSharedPtrPoolDnsAwarePoolManagementError(std::uint_fast64_t /*poolId*/, std::exception&) const noexcept override {}
-            virtual void logSharedPtrPoolDnsAwarePoolManagementError(std::uint_fast64_t /*poolId*/) const noexcept override {}
+            [[deprecated]] virtual void logSharedPtrPoolDnsAwarePoolManagementCycleStart(std::uint_fast64_t /*poolId*/) const noexcept override {}
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/logging.hpp]: deprecated ma variantu kde muzes dat message, premyslim jestli by stalo za to rict co to nahrazuje - nevyhoda je jen, ze je ta informace duplikovana

File: tests-extended/dnsa_connection_pool.cpp (hunk 3 above)
Comment (AdamStepan) [reply tests-extended/dnsa_connection_pool.cpp]: ACK

File: tests-extended/dnsa_connection_pool.cpp (hunk 2 above)
Comment (AdamStepan) [reply tests-extended/dnsa_connection_pool.cpp]: Nepotrebuje, snazim se byt konzistentni. BTW, pouziti {} je preferovane podle  [cpp core guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es23-prefer-the--initializer-syntax)

File: tests/db_access/test_utils.hpp (hunk 5 above)
Comment (AdamStepan) [reply tests/db_access/test_utils.hpp]: Jsem si jisty ze musi byt inicializovany pred pouzitim, nevim jak presne se kompilator zachova kdyz ho nezinicializuju explicitne, IMO "explicit is better than implicit"

File: include/superior_mysqlpp/logging.hpp (hunk 6 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/logging.hpp:655]: Je to dobry napad, ale myslim si ze je to out-of-scope, bylo by to nutne zmenit u vsech metod v loggerech, coz se nevztahuje k tomuhle tasku. BTW, nevis jaky duvod ma pouziti `'\n' << std::flush` misto `std::endl`?

File: include/superior_mysqlpp/logging.hpp (hunk 7 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/logging.hpp]: Taky jsem o tom premyslel, IMO to nestoji za to, muzem se o tom pobavit

File: tests-extended/makefile (hunk 4 above)
Comment (OpatrilPeter) [reply tests-extended/makefile:4]: As Adam said, XXX is a pattern representing something that ought to be fixed, but cannot be changed right now. So OK from me.

File: tests/db_access/test_utils.hpp (hunk 5 above)
Comment (OpatrilPeter) [reply tests/db_access/test_utils.hpp]: Considering that buffer is actually required within the constructor, I agree here with explicitness.

File: include/superior_mysqlpp/logging.hpp (hunk 6 above)
Comment (OpatrilPeter) [reply include/superior_mysqlpp/logging.hpp:655]: I don't see any good reason.

Would you make a ticket for future refactoring of that out-of-scope feature?

File: include/superior_mysqlpp/logging.hpp (hunk 7 above)
Comment (OpatrilPeter) [reply include/superior_mysqlpp/logging.hpp]: I'd at least add a single comment above them then - again, explicit is better then implicit.

Comment (AdamStepan) [reply include/superior_mysqlpp/logging.hpp]: ACK

File: include/superior_mysqlpp/logging.hpp (hunk 6 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/logging.hpp:655]: Yep

File: tests-extended/dnsa_connection_pool.cpp [hunk 8]
```diff
@@ -192,5 +192,46 @@ go_bandit([]() {
...
             AssertThat(pool.poolState().available < min_spare_connections, IsTrue());
         });
+
+        it("can log dns error", [&]() {
+
+            auto &settings = getSettingsRef();
+            std::string hostname{"definitiely.not.resolvable.hostname"}, streamRedirectionResult{};
+
+
+                StreamRedirect<std::ostream> redirection(std::cerr);
```
Comment (OpatrilPeter) [inline tests-extended/dnsa_connection_pool.cpp]: Bad indentation.

File: tests-extended/dnsa_connection_pool.cpp [hunk 9]
```diff
@@ -192,5 +192,46 @@ go_bandit([]() {
             std::this_thread::sleep_for(10ms);
             AssertThat(pool.poolState().available < min_spare_connections, IsTrue());
         });
+
+        it("can log dns error", [&]() {
+
+            auto &settings = getSettingsRef();
+            std::string hostname{"definitiely.not.resolvable.hostname"}, streamRedirectionResult{};
```
Comment (OpatrilPeter) [inline tests-extended/dnsa_connection_pool.cpp]: I see no reason to have streamRedirectionResult here - you can declare it where it's actually used.

File: tests-extended/dnsa_connection_pool.cpp [hunk 10]
```diff
@@ -192,5 +192,46 @@ go_bandit([]() {
...
+                                );
+                            });
+                    },
+                    hostname,
+                    std::make_shared<Loggers::Full>()
+                );
+
+                // ensure that at least one log method is called
+                connectionPool.startDnsAwarePoolManagement();
+                connectionPool.stopDnsAwarePoolManagement();
```
Comment (OpatrilPeter) [inline tests-extended/dnsa_connection_pool.cpp]: Is this line really necessary? I believe it is connection pool's responsibility to stop the management (in the destructor).

File: tests/db_access/test_utils.hpp [hunk 11]
```diff
@@ -7,7 +7,34 @@
...
+        }
+
+        ~StreamRedirect() {
+            restore();
+        }
+
+    private:
+        std::stringstream buffer;
+        T &originalStream;
+        std::streambuf *originalBuffer;
```
Comment (OpatrilPeter) [inline tests/db_access/test_utils.hpp:40]: We can simplify the logic by not storing `originalBuffer`, as each istream already can retrieve a pointer to its internal buffer.
See https://stackoverflow.com/questions/49591500/difference-between-internal-vs-associated-stream-buffer

Note that if the passed stream would have associated buffer different than internal one, this simplification would not work, but we may not care about this case.

File: tests/db_access/test_utils.hpp [hunk 12]
```diff
@@ -7,7 +7,34 @@
 
 #include <chrono>
 #include <thread>
+#include <iostream>
+#include <streambuf>
+#include <string>
+#include <sstream>
 
+template <typename T>
+class StreamRedirect {
```
Comment (OpatrilPeter) [inline tests/db_access/test_utils.hpp]: Please consider more indicative name.

**Review by OpatrilPeter — CHANGES_REQUESTED** [review]

File: tests-extended/dnsa_connection_pool.cpp (hunk 8 above)
Comment (AdamStepan) [reply tests-extended/dnsa_connection_pool.cpp]: ACK

File: tests-extended/dnsa_connection_pool.cpp (hunk 9 above)
Comment (AdamStepan) [reply tests-extended/dnsa_connection_pool.cpp]: ACK

File: tests/db_access/test_utils.hpp (hunk 11 above)
Comment (OpatrilPeter) [reply tests/db_access/test_utils.hpp:40]: We decided that it isn't worth the change, instead enhance commentary of current solution.

**Review by OpatrilPeter — APPROVED** [review]

**Review by majcl — APPROVED** [review]

**Pull request merged**